rmgpy/data/kinetics/family.py

```diff
@@ -2064,30 +2064,35 @@ def getLabeledReactantsAndProducts(self, reactants, products):
         elif len(reactants0) == 2:
             moleculeA = reactants0[0]
             moleculeB = reactants0[1]
+            # get mappings in forward direction
             mappingsA = self.__matchReactantToTemplate(moleculeA, template.reactants[0])
             mappingsB = self.__matchReactantToTemplate(moleculeB, template.reactants[1])
+            mappingPairs = list(itertools.product(mappingsA,mappingsB))
+            # get mappings in the reverse direction
```

Comment:
This method does not check both orderings because it is called by `addAtomLabelsForReaction`, which does check both orderings.

I would suggest seeing if `addAtomLabelsForReaction` is appropriate for your use case, or also modifying that method to account for your changes here.

Reply:
Thanks for brining this up. I wasn't aware that `addAtomLabelsForReaction` accounted for reverse order. The current issue with using `addAtomLabelsForReaction` for my application is that it is written to require reactants and products as molecule objects and outputs a reaction object with species objects (though without resonance structures).  In my application, I start with a reaction with species objects containing possibly resonance structures, and I need to obtain labeled molecule objects for the reactants. It seems like I would have to some finagling of recreating a bunch of reaction objects to input into `addAtomLabelsForReaction` and then extract out the molecules from their respective Species objects. Though I am fine making this correction (or any others) if that would be preferable.

I guess my thoughts on the code currently:

1. Nothing within `getLabeledReactantsAndProducts` seems to imply that the intended outcome is to throw an error if the order of reactants is wrong. If we want it to throw an error in these cases, we should probably make it a private method so users like me don't go trying to use it unintentionally....though making it private anyways may be a good idea.
2. `addAtomLabelsForReaction` seems to do multiple things in one go, which is not good usage of object oriented design (and this is why it doesn't really work well for my application). The conversion to Species objects seems outside the scope of the method and removing this extraneous part could make the method more applicable for other uses.
3. I couldn't find any usage for `addAtomLabelsForReaction` in any of the RMG-Py or RMG-website repositories (the same is true for `getLabeledReactantsAndProducts`). Neither method is tested. I think another alternative to deal with  `addAtomLabelsForReaction` is to deprecate it for removal in RMG 2.2, and only support `getLabeledReactantsAndProducts`.
4. Having only one public method, either `getLabeledReactantsAndProducts` or a more robust and single-tasked version of `addAtomLabelsForReaction` seems like a code improvement to me.

Let me know what you think the best path forward is, and I can implement it.

Reply:
These two methods were written for `convertKineticsLibraryToTrainingReactions.ipynb` in RMG-database. I was looking into them for some fixes to the ipynb, and found that `addAtomLabelsForReaction` broke with reactions with Species objects , as you mentioned. I had made a temporary hack fix by having the method take the first molecule of each of the inputted species, which would then be converted back to a species at the end. I agree that both methods need to be reworked to accept and return reactions with Species objects.

I've written a basic unit test for `addAtomLabelsForReaction`, located [here](https://github.com/mliu49/RMG-Py/commit/e7cbcba9a991c8dad1c22edcb6516247941d24cc), which you can use or expand on. I think you have a lot of flexibility in however you want to remake these two methods, since they're only used in that one notebook. It would be great if you could condense it into a single method. I assumed that the reason it was broken into two was to reduce code duplication.

Reply:
Thanks for the advice. I think it really helped improve the code. I moved over `ensure_species` to make it a function of Reaction objects. I then used that method to make `addAtomLabelsForReaction` much more flexible. It can accept either species- or molecule-based reaction objects and then can output either species- or molecule-based reaction objects. I left the checking-order thing that sparked the discussion in `getLabeledReactantsAndProducts` since it seems more like the intended behavior of the method. 

In the limited case `addAtomLabelsForReaction` was originally used for(input molecule-based reaction object and get out species-based reaction object), the only change to the method is that resonance structures are now output, whereas they were not before. 

The only think I can think of that is still left to be decided is whether we should make `getLabeledReactantsAndProducts` private. I am ambivalent either way (now that `addAtomLabelsForReaction` functions more robustly)